presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -812,7 +813,52 @@ public void mapExpression(VariableReferenceExpression from, VariableReferenceExp
             expressions.put(from, to);
         }
 
-        public void addPlan(PlanNode plan, CanonicalPlan canonicalPlan)
+        private void addPlan(PlanNode plan, CanonicalPlan canonicalPlan)
+        {
+            if (!plan.getStatsEquivalentPlanNode().isPresent()) {
+                addPlanInternal(plan, canonicalPlan);
+                return;
+            }
+            PlanNode statsEquivalentPlanNode = plan.getStatsEquivalentPlanNode().get();
+            // When limits are involved, we can only know canonicalized plans after topmost limit has been canonicalized.
+            // Once we are at topmost limit, we cache canonicalized plans for all sub-plans.
+            if (plan instanceof LimitNode) {
```

Comment:
[major]: how likely we can move this logic to `visitLimit`? `addPlan` is such a generic method being used by every visit method.